### PR TITLE
[JENKINS-68981] Export `hudson.model.Run#isInProgress`

### PR DESCRIPTION
**Pull request by jmdesprez**
See [JENKINS-68981](https://issues.jenkins.io/browse/JENKINS-68981).

Tested manually running [discardBuilds](https://github.com/jenkinsci/acceptance-test-harness/blob/1b1d06e647a807d1ef541facc555abc7b8dc5b08/src/test/java/core/FreestyleJobTest.java#L169) ATH test. Please let me know if an automated test is needed.

### Proposed changelog entries

* Export `hudson.model.Run#isInProgress` to the `json` API.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6829"></a>

### CODE REVIEW

**Comment by daniel-beck** [comment]
The justification seems weak. What other use cases outside of ATH exist for this?

**Comment by jtnord** [comment]
> What other use cases outside of ATH exist for this?

Any API user that wants to know a build has completed.  Currently the only way to know this for sure is to check for the existence of a spinner in the console HTML output by scraping HTML that is likely to be very brittle and is not API friendly (`FINISHED` is printed before the build is `COMPLETE` and listeners have been called!)

**Comment by jtnord** [comment]
Alternatively we could change `isBuilding` to actually return `true` until the build is `COMPLETE` (this opens up the question of when is a Run building?) and is probably more likely to break something that is expecting the current behaviour?

The Javadoc for `isBuilding` - is stating that it should return `true` when the build has not "completed" - yet for `State,POST_PRODUCTION` state this returns `false` and is contrary to the javadoc...

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/model/Run.java
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java:542]: Is the change of modifier actually required? This will be incompatible (at least source, perhaps binary) for Pipeline: https://github.com/jenkinsci/workflow-job-plugin/blob/8def1a5f3944cb5d68ceb955b541a83897b13f98/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L780-L781

**Review by jglick — COMMENTED** [review]
Using the real `inProgress` field is indeed mandatory for correctness. See the special implementation in `WorkflowRun`.

There should be a matching PR to `acceptance-test-harness` demonstrating use of the new API.

**Review by jglick — CHANGES_REQUESTED** [review]
This is not ready.
* https://github.com/jenkinsci/acceptance-test-harness/pull/860 should actually use the new API.
* Should first **release** an update
```diff
diff --git src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java
index fbc5738..f809884 100644
--- src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java
+++ src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java
@@ -778,7 +778,7 @@ public final class WorkflowRun extends Run<WorkflowJob,WorkflowRun> implements F
     }
 
     @Exported
-    @Override protected boolean isInProgress() {
+    @Override public boolean isInProgress() {
         if (Boolean.TRUE.equals(completed)) {  // Has a persisted completion state
             return false;
         } else {
```
* Preferably, fix the annotation process in `stapler` to signal an error if you attempt to put `@Exported` on something which is not `public`.

File: core/src/main/java/hudson/model/Run.java
Comment (jmdesprez) [reply core/src/main/java/hudson/model/Run.java:542]: Yes it is required :/

**Comment by jmdesprez** [comment]
Related PR: https://github.com/jenkinsci/workflow-job-plugin/pull/278, ping @jglick

**Comment by jglick** [comment]
**Proposed upgrade guidelines** will need to mention that you must first upgrade `workflow-job` (if installed), or stuff related to Pipeline builds may break. (I guess—test it!)

**Review by jglick — APPROVED** [review]
I _think_ this will now be OK.

https://github.com/jenkinsci/stapler/blob/03ae7bf98dbe8c312eb727fd3d8c149a8e55087c/core/src/main/java/org/kohsuke/stapler/jsr269/ExportedBeanAnnotationProcessor.java#L74 looks like the place to fail the build when someone tries to write

```java
@Exported
protected SomeType getSomeProperty() {
```

since https://github.com/jenkinsci/stapler/blob/e4c10835694bc8f5db30ffa1c15c50469c60acd8/core/src/main/java/org/kohsuke/stapler/export/Model.java#L98 does appear to only consider `public` methods.

https://github.com/jenkinsci/stapler/blob/e4c10835694bc8f5db30ffa1c15c50469c60acd8/core/src/main/java/org/kohsuke/stapler/export/Model.java#L103-L104 has some runtime warnings which should ideally be in the annotation processor as well.

https://github.com/jenkinsci/stapler/blob/e4c10835694bc8f5db30ffa1c15c50469c60acd8/core/src/test/java/org/kohsuke/stapler/jsr269/ExportedBeanAnnotationProcessorTest.java currently has no tests for error conditions (nonempty `Compilation.diagnostics`).

**Review by olamy — APPROVED** [review]

**Comment by NotMyFault** [comment]
This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
